### ISSUE TITLE
Constant evaluation failure reports an error on each use-site

### ISSUE THREAD

**Issue by retep998** [rust-lang/rust#32842]
In the past few days `winapi` was hit with a deluge of this warning (266 of them to be precise): `warning: constant evaluation error: attempted to subtract with overflow. This will become a HARD ERROR in the future, #[warn(const_err)] on by default`

I can't say I'm too thrilled as `winapi` makes use of this pattern quite a bit in defining constants: `pub const LVN_DELETEALLITEMS: ::UINT = LVN_FIRST - 4;`

The workaround is kinda ugly: `pub const LVN_DELETEALLITEMS: ::UINT = (LVN_FIRST as ::INT - 4) as ::UINT;`

**Comment by nikomatsakis** [comment]
cc me, @oli-obk, @rust-lang/lang

**Comment by nikomatsakis** [comment]
@retep998 presumably `LVN_FIRST` is `0` here?

**Comment by retep998** [comment]
`pub const LVN_FIRST: UINT = 0 - 100`

**Comment by nikomatsakis** [comment]
@retep998 points out on IRC that we really ought to report exactly one overflow here -- for `LVN_FIRST`. I tend to agree. More importantly, it ought to be that @retep998 only needs to change that one definition, not all of them.

@oli-obk, do you agree? How hard do you think it would be to achieve that?

**Comment by retep998** [comment]
By swapping out `pub const LVN_FIRST: ::UINT = 0 - 100;` for `pub const LVN_FIRST: ::UINT = -100i32 as ::UINT;` it does eliminate the warnings on the constants that depend on `LVN_FIRST`.
There are only 44 hits for `0 -` so the actual work done to fix it on my end is somewhat less than 266. 
Do note that one of those constants on which a lot of other constants subtract from is `pub const NM_FIRST: ::UINT = 0;` so I still have to do the really ugly workaround for those constants.

**Comment by arielb1** [comment]
@retep998 

Of course, the attempted wraparound subtraction is, of course, erroneous and should not compile. We should not emit errors on use-sites, however.

**Comment by arielb1** [comment]
Example:

```Rust
const C: usize = 1/0;
fn main() {
    let _ = C;
    let _ = C;
    let _ = C;
    let _ = C;
}
```

This gives 4 errors on all versions:
```
<anon>:1:18: 1:21 warning: attempted to divide by zero, #[warn(const_err)] on by default
<anon>:1 const C: usize = 1/0;
                          ^~~
<anon>:1:18: 1:21 error: attempted to divide by zero
<anon>:1 const C: usize = 1/0;
                          ^~~
<anon>:1:18: 1:21 error: attempted to divide by zero
<anon>:1 const C: usize = 1/0;
                          ^~~
<anon>:1:18: 1:21 error: attempted to divide by zero
<anon>:1 const C: usize = 1/0;
                          ^~~
<anon>:1:18: 1:21 error: attempted to divide by zero
<anon>:1 const C: usize = 1/0;
```

**Comment by oli-obk** [comment]
These are two distinct issues. One is the new lint and easy to fix. I'll have a PR next week. The other is in the old trans code, and impossible to fix without breaking safety until the lint is a hard error. Or we just wait for MIR to surpass the old trans

**Comment by nikomatsakis** [comment]
@oli-obk sorry, which is the issue that is impossible to fix specifically?

**Comment by oli-obk** [comment]
It's not impossible to fix, it just can't be fixed without a cache for errors and/or constants.

The issue I meant is that an error is reported additionally at a use site, when using a broken constant in a normal expression. This occurs, because trans/consts doesn't cache constants, but instead evaluates them through llvm at every use site (in contrast to statics, which are only evaluated once). Note that this error reporting happens at trans time, so if the const_err lint is set to deny, we never get there.

**Comment by nikomatsakis** [comment]
@oli-obk by error do you mean lint warning?

**Comment by oli-obk** [comment]
No. We currently report const eval errors in a few locations, with differing criticality:

1. the const_err lint is triggered in the `check_const` pass on all constants and statics
2. the const_err lint is triggered in the `check_const` pass on division by zero for all expressions that can be constant evaluated
3. during trans, if a const eval error occurs outside a constant's expression, a warning is triggered and normal trans is used
4. during trans, if a const eval error occurs in a constant's expression, an error is triggered and compilation is aborted at some point.

The code shown by @arielb1 triggers 2 and 4 on stable and 1, 2 and 4 on nightly (and has been doing so since [rfc1229](https://github.com/rust-lang/rfcs/blob/master/text/1229-compile-time-asserts.md)).

We can probably eliminate 2 from triggering inside constants

Since 1 is just a lint, it doesn't stop compilation and so the expression is still compiled. Trans will need to report an error, because otherwise we'd end up miscompiling the constant. We could make sure that it only shows up once, by caching translated constants (and thus translating every constant only once). With the lint enabled we'd still report the const eval error twice though, once in the lint, and once at the first use site.

Turning the lint into a hard error will eliminate all of these duplicates, because compilation can never get that far. This would also greatly simplify trans/consts, because we can simply `bug!` out wherever we reported errors before.
